File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        
        vector<int> res(temperatures.size(), 0);
        
        stack<pair<int, int>> st;
        
        for(int i=0;i<temperatures.size();i++){
            
            int t = temperatures[i];
            
            while(!st.empty() && t > st.top().first){
                
                int stackTop = st.top().first;
                int stackIndex = st.top().second;
                
                st.pop();
                
                res[stackIndex] = i - stackIndex;
            }
            
            st.push({t, i});
        }
        return res;
    }
};

/*
int main() {
    Solution solution;
    vector<int> temperatures = {73, 74, 75, 71, 69, 72, 76, 73}; // You can change this list to test with different temperatures
    vector<int> result = solution.dailyTemperatures(temperatures);

    // Print the result, which contains the number of days to wait for a warmer temperature
    for (int i : result) {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}
*/

File: 0739-daily-temperatures/0739-daily-temperatures.py
## space: o(n)
## time: o(n)

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        
        ##initialize a result array of length same as input array with default values as '0'
        
        res = [0] * len(temperatures)
        
        
        ##initializw a stack with pair values
        
        stack = []
        ## pair : [temp, index]
        
        ## itetate and keep track of both the value and index in the input array
        
        for i,t in enumerate(temperatures):
            
            ## check if the stack is not empty and temperature value is greater than stack top value
            
            while stack and t > stack[-1][0]:
                
                ## pop those value and index from the stack
                stackTop, stackIndex = stack.pop()
                
                ## append the difference into result
                
                res[stackIndex] = i - stackIndex
            stack.append((t, i))
                
        return res         
                
        